packages/headless/src/controllers/url-manager/headless-url-manager.ts

```diff
@@ -77,25 +72,42 @@ export function buildUrlManager(
     throw loadReducerError;
   }
 
-  const {dispatch} = engine;
-  const controller = buildController(engine);
-
   validateInitialState(
     engine,
     initialStateSchema,
     props.initialState,
     'buildUrlManager'
   );
 
+  const completeParameters = (fragment: string) => ({
+    ...initialSearchParameterSelector(engine.state),
+    ...convertFragmentToParameters(fragment),
+  });
+
+  const controller = buildController(engine);
+  const {dispatch} = engine;
+  let previousFragment = decodeFragment(props.initialState.fragment);
   const searchParameterManager = buildSearchParameterManager(engine, {
     initialState: {
-      parameters: fragmentActiveSearchParameters(props.initialState.fragment),
+      parameters: convertFragmentToParameters(previousFragment),
     },
   });
 
   return {
     ...controller,
 
+    subscribe(listener: () => void) {
+      const strictListener = () => {
+        const newFragment = this.state.fragment;
+        if (!compareFragments(previousFragment, newFragment)) {
+          previousFragment = newFragment;
+          listener();
+        }
+      };
+      strictListener();
```

Comment:
Using a "strictListener" on the first subscribe here, we don't want the controller's user to pushState if it's the same fragment. In the case where it changes between it's initialization and the subscribe, it will still work.